doc: add description about CMake options (#1403)
groonga/groonga: doc/source/install/cmake.rst

.. -*- rst -*-

> 

Reviewer: We can remove this because we use `.rst` as extension.
(We used to use `.txt`.)

Author: I see.

Author: Fixed.

How to build Groonga with CMake
===============================

This document describes how to build Groonga from source with CMake.

To get more details about installing Groonga from the source with CMake on a specific environment,
find the document for the specific environment from :doc:`/install`.

Install depended software
-------------------------

> Here is lists of depended software for GNU/Linux, UNIX and Windows.

Reviewer: ```suggestion
Here is a list of depended software for GNU/Linux, UNIX and Windows.
```

or

```suggestion
Here is depended software for GNU/Linux, UNIX and Windows.
```

Author: Fixed.

> TODO

Author: We should check and add all required software.
I migrated what was originally a TODO in `unix_cmake.rst`, so it's still a TODO.

https://github.com/HashidaTKS/groonga/blob/0566aceb53825ac9f964adbd490e8a5babb66ae2/doc/source/contribution/development/build/unix_cmake.rst#L21

GNU/Linux or UNIX
+++++++++++++++++

  * `CMake <http://www.cmake.org/>`_

Windows
+++++++

  * `Microsoft Visual Studio Community
    <https://visualstudio.microsoft.com/vs/community/>`_
  * `CMake <http://www.cmake.org/>`_

Download source
---------------

You can get the latest source from `packages.groonga.org <https://packages.groonga.org/source/groonga>`_.

GNU/Linux or UNIX
+++++++++++++++++

.. code-block:: console

   $ wget https://packages.groonga.org/source/groonga/groonga-12.0.7.tar.gz
   $ tar xvzf groonga-12.0.7.tar.gz

Windows
+++++++

Download the latest zipped source from packages.groonga.org.

  * https://packages.groonga.org/source/groonga/groonga-12.0.7.zip

Then extract it.

.. _cmake_run:

Run ``cmake``
-------------

> You need to create ``Makefile`` for your environment.

Reviewer: ```suggestion
You need to generate build files such as ``Makefile`` for your environment.
```

Author: Fixed.

The default is ``OFF``.

You can install Apache Arrow following to `the official installation procedure <https://arrow.apache.org/install/>`_. 

Here is an example how to enable the Apache Arrow support.

.. code-block:: console

   $ cmake . -DGRN_WITH_APACHE_ARROW=ON

.. note::

   If you install Apache Arrow manually, you need to use the :ref:`cmake-options-cmake-prefix-path` option.

.. _cmake-options-cmake-prefix-path:

``-DCMAKE_PREFIX_PATH=PATHS``
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Adds search paths for ``.cmake`` files.

You can specify multiple path separating them with ``:`` on GNU/Linux or UNIX, ``;`` on Windows.

In case of using libraries installed via a package manager, you do not need to specify this 
parameter. It is because ``.cmake`` files for those libraries are in the default search paths of CMake.

In case of using libraries installed in non-system directories such as ``/usr``, you need to specify ``.cmake`` file paths of those libraries by this parameter.

Here is an example how to specify a ``.cmake`` file path for ``/tmp/local/lib/cmake/Arrow/ArrowConfig.cmake`` on GNU/Linux or UNIX.

.. code-block:: console

>    $ cmake . -DCMAKE_PREFIX_PATH="/tmp/local/lib/cmake/Arrow"

Reviewer: ```suggestion
   $ cmake . -DCMAKE_PREFIX_PATH="/tmp/local"
```

Author: Fixed

Here is an example how to specify a ``.cmake`` file path for ``C:\arrow\lib\cmake\Arrow\ArrowConfig.cmake`` on Windows.

.. code-block:: pwsh-session

>    > cmake . -DCMAKE_PREFIX_PATH="C:\arrow\lib\cmake\Arrow"

Reviewer: ```suggestion
   > cmake . -DCMAKE_PREFIX_PATH="C:\arrow"
```

.. _cmake-build-and-install:

Build and install Groonga
-------------------------

Now, you can build Groonga.

GNU/Linux or UNIX
+++++++++++++++++

> you can use ``make``.

Reviewer: ```suggestion
You can use ``make``.
```

Here is a command line to build and install Groonga by ``make``.

.. code-block:: console

>   $ make -j8
>   $ sudo make install

Reviewer: The following will work on GNU/Linux, FreeBSD and macOS:

```suggestion
   $ make -j$(nproc || PATH="/sbin:$PATH" sysctl -n hw.ncpu)
   $ sudo make install
```

``-j8`` decreases build time. It enables parallel build. If you have 8
or more CPU cores, you can increase ``8`` to decreases more build

> time.

Reviewer: We can remove this description if we use `$(nproc)` or something because it choose the suitable number of jobs.

You can just see only warning and error messages by ``> /dev/null``.

.. code-block:: console

>   $ make -j8 > /dev/null

Reviewer: ```suggestion
   $ make -j8 > /dev/null
```

Windows
+++++++

You can use Visual Studio or ``cmake --build``.

> Here is a command line to build and install Groonga by ``cmake --build``.

Reviewer: FYI: We can use `cmake --build` for `Unix Makefiles` too. We may unify build step description in the future.

Reviewer: We can use the `CMAKE_BUILD_PARALLEL_LEVEL` environment variable for parallel build with ``cmake --build`: https://cmake.org/cmake/help/latest/envvar/CMAKE_BUILD_PARALLEL_LEVEL.html

Author: Ok, I will add description about `CMAKE_BUILD_PARALLEL_LEVEL` and `--parallel [<jobs>], -j [<jobs>]` (https://cmake.org/cmake/help/latest/manual/cmake.1.html#build-tool-mode) in the future.
